waku/v2/protocol/peer_exchange/shard_lru.go

```diff
@@ -0,0 +1,167 @@
+func (l *shardLRU) remove(node *enode.Node) {
+	elements := l.idToNode[node.ID()]
+	for _, element := range elements {
+		key := element.Value.(nodeWithShardInfo).key
+		l.shardNodes[key].Remove(element)
+	}
+	delete(l.idToNode, node.ID())
+}
+
+// if a node is removed for a list, remove it from idToNode too
+func (l *shardLRU) removeFromIdToNode(ele *list.Element) {
+	nodeID := ele.Value.(nodeWithShardInfo).node.ID()
+	for ind, entries := range l.idToNode[nodeID] {
+		if entries == ele {
+			l.idToNode[nodeID] = append(l.idToNode[nodeID][:ind], l.idToNode[nodeID][ind+1:]...)
+			break
+		}
+	}
+	if len(l.idToNode[nodeID]) == 0 {
+		delete(l.idToNode, nodeID)
+	}
+}
+
+func nodeToRelayShard(node *enode.Node) *protocol.RelayShards {
+	shard, err := wenr.RelaySharding(node.Record())
+	if shard == nil || err != nil { // if no shard info, then add to node to Cluster 0, Index 0
+		shard = &protocol.RelayShards{
+			ClusterID: 0,
+			ShardIDs:  []uint16{0},
+		}
+	}
+	return shard
+}
+
+// time complexity: O(new number of indexes in node's shard)
+func (l *shardLRU) add(node *enode.Node) {
+	shard := nodeToRelayShard(node)
+	elements := []*list.Element{}
+	for _, index := range shard.ShardIDs {
+		key := ShardInfo{
+			shard.ClusterID,
+			index,
+		}
+		if l.shardNodes[key] == nil {
+			l.shardNodes[key] = list.New()
+		}
+		if l.shardNodes[key].Len() >= l.size {
+			oldest := l.shardNodes[key].Back()
+			l.removeFromIdToNode(oldest)
+			l.shardNodes[key].Remove(oldest)
+		}
+		entry := l.shardNodes[key].PushFront(nodeWithShardInfo{
+			key:  key,
+			node: node,
+		})
+		elements = append(elements, entry)
+
+	}
+	l.idToNode[node.ID()] = elements
+}
+
+// this will be called when the seq number of node is more than the one in cache
+func (l *shardLRU) Add(node *enode.Node) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	// removing bcz previous node might be subscribed to different shards, we need to remove node from those shards
+	l.remove(node)
+	l.add(node)
+}
+
+// clusterIndex is nil when peers for no specific shard are requested
+func (l *shardLRU) GetRandomNodes(clusterIndex *ShardInfo, neededPeers int) (nodes []*enode.Node) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	availablePeers := l.len(clusterIndex)
+	if availablePeers < neededPeers {
+		neededPeers = availablePeers
+	}
+	// if clusterIndex is nil, then return all nodes
+	var elements []*list.Element
+	if clusterIndex == nil {
+		elements = make([]*list.Element, 0, len(l.idToNode))
+		for _, entries := range l.idToNode {
+			elements = append(elements, entries[0])
+		}
+	} else if entries := l.shardNodes[*clusterIndex]; entries != nil && entries.Len() != 0 {
+		elements = make([]*list.Element, 0, entries.Len())
+		for ent := entries.Back(); ent != nil; ent = ent.Prev() {
+			elements = append(elements, ent)
+		}
+	}
+	utils.Logger().Info(fmt.Sprintf("%d", len(elements)))
+	indexes := l.rng.Perm(len(elements))[0:neededPeers]
+	for _, ind := range indexes {
```

Comment:
Optimization and can be taken up later: I am wondering if rather than deleting and adding nodes everytime there is a query to keep an indication of recently used nodes, would it be better to just update last fetched timestamp for node that has been returned. 

Only when cache is full and new nodes are to be added, then this timestamp can be used to determine which is least used and remove those nodes. 

Reason being, for a peer that is acting as a peer-exchange node, the number of queries to get nodes may be more than nodes getting added dynamically. But, this can always be relooked at a later stage based on node-churn and behavior in the network.